### PR TITLE
[FIX] sale_coupon: Apply taxes to fixed_price discount

### PR DESCRIPTION
**Pull request by fdamhaut**
Step to reproduce:
- Create a promotion with fixed price discount
- Create a SO where the promotion can apply with taxes on the
products
- Trigger the computation of the taxes

Current behaviour:
- No taxes are applied on the fixed price discount

Behaviour after PR:
- Taxes are applied on the fixed price based on their
proportion to the total amount.

opw-2806780

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Williambraecky — COMMENTED** [review]
Hello 👋 
I feel like @rdeodoo is more inclined to review this properly.

Overall the change LGTM but I did not review in depth.

File: addons/sale_coupon/tests/test_program_numbers.py [hunk 1]
```diff
@@ -705,9 +706,9 @@ def test_coupon_and_coupon_discount_fixed_amount_tax_excl(self):
             }).process_coupon()
         order.recompute_coupon_lines()
         #TODO fix numbers
-        self.assertEqual(order.amount_tax, 13.5)
+        self.assertEqual(order.amount_tax, 0)
```
Comment (rdeodoo) [inline addons/sale_coupon/tests/test_program_numbers.py:716]: > #TODO fix numbers

Well, you just did, remove that comment ;-)

File: addons/sale_coupon/tests/test_program_numbers.py [hunk 2]
```diff
@@ -687,9 +688,9 @@ def test_coupon_and_coupon_discount_fixed_amount_tax_excl(self):
         # - a fixed amount (greater than remaining amount to pay) +
         # - discount amount
         # And user should be able to swap the promotion order with a meaningful result.
-        self.assertEqual(order.amount_tax, 13.5)
+        self.assertEqual(order.amount_tax, 0)
```
Comment (rdeodoo) [inline addons/sale_coupon/tests/test_program_numbers.py:698]: > #TODO fix numbers
        # Need an in-depth inspection on the behavior with
        # - multiple product with different VAT +
        # - a fixed amount (greater than remaining amount to pay) +
        # - discount amount
        # And user should be able to swap the promotion order with a meaningful result.

Your fix is supposed to handle all that, now that you are tackling this issue that has been postponed for a long time.
I think every part of that comment is true, if we don't want to have a bad risk/reward ratio with this fix.

I actually wrote my review message before reading this part of the code which actually.. Explains the exact same thing it seems, making it even more relevant.

File: addons/sale_coupon/models/sale_order.py [hunk 3]
```diff
@@ -141,20 +141,49 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
...
+        lines = lines.filtered('price_reduce')
+        lines_total_amount = sum(lines.filtered(lambda l: l.price_total > 0).mapped('price_total'))
+        reward_lines = {}
+
+        for line in lines:
+
+            taxes = line.tax_id
+
+            line_price = line.price_reduce * line.product_uom_qty
+            # Next line is equivalent to computing the % of taxes included price then getting the related sale_price
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:163]: Never go over 80 char / line in comments and docstring. https://github.com/odoo/odoo/pull/83508#discussion_r796442104 might be useful

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Hello and thanks for the PR.

Hmm, seems like this is fixing the issue I highlighted at https://github.com/odoo/odoo/pull/85955#issuecomment-1075100394
As at that time it was not the issue mentioned by the client but:
1. Something I found during tests
2. Handled in the master refactoring

It was decided to ignore that part.

I will have a look tomorrow, this is just a quick one scrolling the files, but could you please try to reproduce the case I was explaining in the link just mentioned, with your PR?

File: addons/sale_coupon/tests/test_program_numbers.py (hunk 2 above)
Comment (fdamhaut) [reply addons/sale_coupon/tests/test_program_numbers.py:698]: > # - multiple product with different VAT +
This is the test that have been added

> # - a fixed amount (greater than remaining amount to pay) + discount amount
This is the test case that was already present (the one you reviewed on), it's the reason for the non-test change in the second commit

> # And user should be able to swap the promotion order with a meaningful result.
Currently promotion are applied in the order of their application, which might result in different SO (if e.g. they have a total discount value of > 100% it will always be the later who get reduced)
But i'll add a test to see if applying 2 coupon/promo in different order result in similar SO if their value < product value

**Comment by Williambraecky** [comment]
Hey 👋 

Since I was pinged I did test something I was wondering about.

Capture d’écran 2022-04-27 à 15 10 07

Do you guys think this result is correct?
The reward lines are quite explicit in what they do.

I am not sure what is expected in stable for this.

**Comment by fdamhaut** [comment]
> Hey wave
> 
> Since I was pinged I did test something I was wondering about.
> 
> Capture d’écran 2022-04-27 à 15 10 07
> 
> Do you guys think this result is correct? The reward lines are quite explicit in what they do.
> 
> I am not sure what is expected in stable for this.

I guess it'll depend on the order in which we set the promo, here i guess that the 5$ are applied first then the 100% discount (because we see that this discount is not complete).
We might want to introduce a rule in the order of the application of promo (on_product before on_order)

**Comment by Williambraecky** [comment]
@fdamhaut I am afraid you went a little too quick with it 😅 

I remember a discussion with @rdeodoo were it was decided in stable for discounts to only take the initial value into account or something along those lines.

This is actually a pretty big change now. And I believe the example I sent previously was actually correct. It's just the lines that are in a weird order.

**Comment by rdeodoo** [comment]
@Williambraecky Are you implying something seems weird or the opposite?

I had a quick look about the case I reported https://github.com/odoo/odoo/pull/85955#issuecomment-1075100394 and it seems to be perfectly handled here:

75% discount + 100$ fixed amount
image

Same but applied the other way around (not sure how I ended up with the reverse apply order, somehow by archiving/unarchiving the promotions between recomputations)
image


Only the 75% discount
image

Only the 100$ fixed discount
image

For the record, without any promotion:
image

Before digging into a technical / code review, do we all agree the results provided by this PR seems to be the expected ones?
AFAIC, it seems to perfectly fixes https://github.com/odoo/odoo/pull/85955#issuecomment-1075128914 and https://github.com/odoo/odoo/pull/85955#issuecomment-1075100394

**Comment by rdeodoo** [comment]
image

For what it's worth, I also checked the eshop/frontend result, and also there everything seems fine.
The 4 discount lines are merged in 2 discount lines so the VAT split is transparent to the user.

As far as the ratio between the taxes and the total is fine (eg, 50% tax on product should result in 50% tax on order after discount, ratio should be the same), everything should be good.

As usual, given the strong test suite we now have in this module, this should be good.

I will have a final in depth look at the code tomorrow and merge it.
(Maybe I will write a test for the exact use case I sent in the screenshot, with usual real number in the `assertEquals`, not a big fan of the fraction, hard to follow whats going ;-))

Thanks for your work already ;-)

File: addons/sale_coupon/tests/test_program_numbers.py [hunk 4]
```diff
@@ -446,11 +447,17 @@ def test_program_numbers_one_discount_line_per_tax(self):
         # Free Pedal Bin       |  5  |   -100.00  | /        | -500.00 | -500.00 |       /
         # Free Large Cabinet   |  3  |   -100.00  | 15% excl | -300.00 | -345.00 |  -45.00
         # 20% on Large Cabinet |  1  |    -80.00  | 15% excl | -80.00  | -92.00  |  -12.00
+        # 10% on no tax        |  1  |    -50.00  | /        | -50.00  | -50.00  |       /
+        # 10% on tax 15% excl  |  1  |    -40.00  | 15% excl | -60.00  | -69.00  |   -9.00
```
Comment (Williambraecky) [inline addons/sale_coupon/tests/test_program_numbers.py:451]: This line is misleading 🙂it's actually a 60$ price_unit.

Side note:
I compared the result with `sale_loyalty` here: https://github.com/odoo/odoo/blob/master/addons/sale_loyalty/tests/test_program_numbers.py#L554-L573

There is a slight change in behaviour so the amounts are slightly different but it seems correct to me.
You are applying -10% on 700+200-300 (free product) = 60$
Where as `sale_loyalty` is doing -10% on 400+200-80 (20% off) = 52$ (note that free products aren't represented as correction lines anymore)

```suggestion
        # 10% on tax 15% excl  |  1  |    -60.00  | 15% excl | -60.00  | -69.00  |   -9.00
```

File: addons/sale_coupon/models/sale_order.py [hunk 5]
```diff
@@ -187,6 +223,7 @@ def _get_reward_values_discount(self, program):
             currently_discounted_amount = 0
             for line in lines:
                 discount_line_amount = min(self._get_reward_values_discount_percentage_per_line(program, line), amount_total - currently_discounted_amount)
+                discount_line_amount = min(self._get_reward_values_discount_percentage_per_tax(program, line.tax_id), discount_line_amount)
```
Comment (Williambraecky) [inline addons/sale_coupon/models/sale_order.py:226]: I am worried about the performance as this could mean potentially `O(n^2n)` if I'm not mistaken.

Though I would expect most orders to not have a lot of order lines and/or discounts applied to them so this might be an invalid concern.

File: addons/sale_coupon/models/sale_order.py [hunk 6]
```diff
@@ -375,13 +414,17 @@ def update_line(order, lines, values):
...
                     # Case 2.
                     if not value_found:
-                        order.write({'order_line': [(0, False, value)]})
+                        lines_to_add += [(0, False, value)]
                 # Case 3.
-                lines_to_remove.unlink()
+                if lines_to_remove:
+                    lines_to_remove.unlink()
+                    order.write({'order_line': lines_to_keep})
+                order.write({'order_line': lines_to_add})
```
Comment (Williambraecky) [inline addons/sale_coupon/models/sale_order.py:427]: Out of scope for this PR but couldn't these be done in a single write? Using `Command.UNLINK`

**Review by Williambraecky — COMMENTED** [review]
I haven't tested a lot but on paper seems good!

File: addons/sale_coupon/models/sale_order.py (hunk 6 above)
Comment (fdamhaut) [reply addons/sale_coupon/models/sale_order.py:427]: I didn't think about it before but I'll test doing it :D

**Comment by fdamhaut** [comment]
@rdeodoo can you have the final look ? (And/or should i make other test)
Since the last review I only change what was pointed out by William

File: addons/sale_coupon/models/sale_order.py [hunk 7]
```diff
@@ -405,14 +414,19 @@ def update_line(order, lines, values):
                             value_found = True
                             # Working on Case 3.
                             lines_to_remove -= line
+                            lines_to_keep += [(0, False, value)]
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:417]: This whole change (with below lines) is "just" for perfs right? Move it in a separate commit please.

File: addons/sale_coupon/models/sale_order.py [hunk 8]
```diff
@@ -405,14 +414,19 @@ def update_line(order, lines, values):
...
                     # Case 2.
                     if not value_found:
                         lines_to_add += [(0, False, value)]
                 # Case 3.
-                lines_to_remove.unlink()
-                order.write({'order_line': lines_to_add})
+                line_update = []
+                if lines_to_remove:
+                    line_update += [(3, lid, 0) for lid in lines_to_remove.ids]
+                    line_update += lines_to_keep
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:427]: Why do you need this line?
`lines_to_keep` could have value that should be deleted and in `lines_to_remove`
```py
# Working on Case 3.
lines_to_remove -= line
lines_to_keep += [(0, False, value)]  # (new line from this PR)
lines_to_remove += update_line(order, line, value)
continue
```

Why keeping a value that might going to be removed?
Either I am tired and this is obvious or this change is a bit strange.

**Review by rdeodoo — CHANGES_REQUESTED** [review]

File: addons/sale_coupon/models/sale_order.py (hunk 6 above)
Comment (rdeodoo) [reply addons/sale_coupon/models/sale_order.py]: Nice catch, could have been part of a separate PR tho, to not noise the commits as those are already quite touchy :)

**Comment by rdeodoo** [comment]
I push forced some of the small nitpicking changes and resolved them

File: addons/sale_coupon/models/sale_order.py [hunk 9]
```diff
@@ -141,20 +141,52 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
...
+                    'is_reward_line': True,
+                    'tax_id': [(4, tax.id, False) for tax in line.tax_id],
+                }
+        return reward_lines.values()
+
     def _get_reward_values_discount(self, program):
         if program.discount_type == 'fixed_amount':
-            taxes = program.discount_line_product_id.taxes_id
-            if self.fiscal_position_id:
-                taxes = self.fiscal_position_id.map_tax(taxes)
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:148]: Just to summarize and confirm.

This is just removed, not adapted. Which seems indeed correct, as it was done for people who added taxes manually on their promotion program product (the one created when you create your promotion).

This is not something rare, probably mostly due to the fact the fixed amount discount lines did not get a tax applied (which you are fixing with this PR).

Eg, https://github.com/odoo-dev/odoo/commit/7633f963066c4962a5d722b7e755e8468b4c2b54

Everything still seems to be working fine for those cases, as the applied tax on that product will just be ignored and the one from the SOL the reward is applied will be considered.
(And that SOL tax is already mapped, so the discount line already received the mapped tax)

File: addons/sale_coupon/models/sale_order.py [hunk 10]
```diff
@@ -141,6 +141,10 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
         discount_amount = line.product_uom_qty * line.price_reduce * (program.discount_percentage / 100)
         return discount_amount
 
+    def _get_reward_values_discount_percentage_per_tax(self, program, taxes):
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:144]: This method is not returning a percentage amount, the name is misleading

**Review by rdeodoo — CHANGES_REQUESTED** [review]
I have weird results, I didn't check in depth why this happen, I just played with some basic cases, please see video and screenshot below.

Basically:
- I first just create a SOL with a tax (200$), and a second SOL with no tax (200$).
- I then applied a 100$ fixed discount with a coupon code and I have 2 lines instead of one (the first SOL can "eat" the fixed discount, no need to split it.
Then in the video:
- With the same 2 initial SOL
- I apply the same coupon
- I then compute the auto applied promotions (2 promotions of -100$ fixed, eg one if you buy an ipad and one if you buy an iphone)
- I then add a 10% discount promotion code on top of that


https://watch.screencastify.com/v/2xqgJY3DzupIq4sccvNE

image

File: addons/sale_coupon/models/sale_order.py [hunk 11]
```diff
@@ -141,20 +141,52 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
...
         return discount_amount
 
+    def _get_reward_values_fixed_amount(self, program):
+
+        discount_amount = - self._get_reward_values_discount_fixed_amount(program)
+
+        lines = self._get_paid_order_lines()
+        # Remove Free Lines
+        lines = lines.filtered('price_reduce')
+        lines_total_amount = sum(lines.filtered(lambda l: l.price_total > 0).mapped('price_total'))
```
[the remark this answers is not in the document]
Comment (fdamhaut) [reply addons/sale_coupon/models/sale_order.py:155]: :eyes: More than being useless, that filter generate wrong value when we already have a negative line.

With the filter : 
image


Without the line : 
image

I will add the example as a testcase.

**Comment by fdamhaut** [comment]
After a fresh look at it, I tough that there might be issue with the fixed taxes (and they are here)
As we compute every taxes if they were % taxes, i'll try to look at how to fix them
image

**Comment by fdamhaut** [comment]
@rdeodoo Do you believe we should complexify the promotion program to be capable to handle part of the fixed tax issue ? (Like the one shown)
We would need to split the lines in 2, 'with and without fixed tax' and then apply different logic on the two with a max discount pro-rated to their total. Which would still have the corner case explained earlier that are 'unsolvable' with current discount line.

I believe that it's better not to have something too complex which would still break and instead say that we cannot fix this for fixed taxes (at least in stable).

**Comment by fdamhaut** [comment]
(And TY for cleaning the commits, I learned recently about fixup and will try to be more mindful of the commit goal)

File: addons/sale_coupon/models/sale_order.py (hunk 8 above)
Comment (fdamhaut) [reply addons/sale_coupon/models/sale_order.py:429]: I don't know which line exactly you were talking about but there was indeed and issue with line_to_keep which might have some lines that need to be removed in case the value has `0` as qty or price_reduced.
+ Added a testcase

**Comment by fdamhaut** [comment]
> I have weird results, I didn't check in depth why this happen, I just played with some basic cases, please see video and screenshot below.
> 
> Basically:
> 
> * I first just create a SOL with a tax (200$), and a second SOL with no tax (200$).
> * I then applied a 100$ fixed discount with a coupon code and I have 2 lines instead of one (the first SOL can "eat" the fixed discount, no need to split it.
>   Then in the video:
> * With the same 2 initial SOL
> * I apply the same coupon
> * I then compute the auto applied promotions (2 promotions of -100$ fixed, eg one if you buy an ipad and one if you buy an iphone)
> * I then add a 10% discount promotion code on top of that
> 
> https://watch.screencastify.com/v/2xqgJY3DzupIq4sccvNE
> 
> image

You have 430$ original price - 3* 100$ discount - 10% of original price(43$) = 87$ 
I don't understand the problem

**Comment by fdamhaut** [comment]
>100+ order line in an order is not very rare, somehow there was some perf opw about coupon and such sale order a few years ago.
Having tons of programs is also quite usual as when people are really using the coupon module, they end up creating a lot of them for everything.
This should probably be a concern if there is a perf hole here

Should not be an issue now, there is 1 call to the function per existing group of taxes which is then updated to avoid recomputing it

**Comment by rdeodoo** [comment]
> After a fresh look at it, I tough that there might be issue with the fixed taxes (and they are here)

Well, you can ignore that part, it is already not working well with the % discounts. That's a limitation of the coupon module in stable. We came a long way with coupon in stable, it is acceptable to not support this. The risk/reward would be awful anyway.

The best you can do is check if it works with the new coupon refactoring and if not, report it.

> You have 430$ original price - 3* 100$ discount - 10% of original price(43$) = 87$
I don't understand the problem

It is not about the total but just about not splitting the discount per tax unless we really need to. We used that in last ressort to fix the accounting and tax reports for the % discounts, in which cases we had to apply the discount on every line + making sense to the customer.

But in this case, why not applying the fixed amount discount on a SOL, then on another one if there is a remaining amount to be discounted and so on. With lines ordered by amount (and by tax).

So if you have a 100$ discount

And those SOL:
- 20$ 10% tax
- 15$ 10% tax
- 110$ no tax

You just create a 100$ discount with no tax

if you have
- 20$ 10% tax
- 15$ 10% tax
- 90$ no tax

There you need to split, 90$ discount with no tax and the remaining amount (10$) with taxes

if you have
- 20$ 10% tax
- 45$ 10% tax
- 45$ 10% tax
- 90$ no tax

You have 90$ discount no tax, 10$ with taxes.

Or rather than sorting by amount and then by taxes, you could group it by tax and order by the amount of those groups.

Eg, instead of the previous example which is finding 90$ as the biggest line but not smart enough to figure there is enough 10% tax SOL to "eat" the reward without splitting it, it could be doing "I have 110$ worth of product with 10% tax and 90$ no tax, so apply the discount first on the 10% SOL):
- 20$ 10% tax
- 45$ 10% tax
- 45$ 10% tax
- 90$ no tax

You have 100$ discount with 10$ with taxes.

So basically, instead of 
image

You end up with either
image

or
image

It is way better for the customer to receive a clear invoice/SO with only one line of discount when we can, rather than splitting it, right?

Hopefully I didn't miss anything ^^

**Comment by fdamhaut** [comment]
@rdeodoo Taxes are not just 'extra' money to pay.
If you have 50$ with 10% tax and 50$ with 20% tax.
As a rule of thumb it mean that you will receive 130$ from selling them and giving back 30$ to the state. (Minus potential cost incurred in creating those goods)

If we step out of the legality of choosing on what the promotion apply (bcs i have 0 idea) choosing the least taxed item would result in the user paying the highest tax % possible (which they would not be happy about).
On the other hand doing that on the product with the highest tax % might be a good for the customer/Odoo user but seems fishy to me.

**Comment by rdeodoo** [comment]
> Taxes are not just 'extra' money to pay.
If you have 50$ with 10% tax and 50$ with 20% tax.
As a rule of thumb it mean that you will receive 130$ from selling them and giving back 30$ to the state. (Minus potential cost incurred in creating those goods)

Of course, that why I initially splitted the discount / line a few years ago, to have a correct tax report: https://github.com/odoo/odoo/pull/85955#issuecomment-1075128914
And this is why the current implementation (in stable) is terrible for fixed amount discount: https://github.com/odoo/enterprise/pull/2224

> If we step out of the legality of choosing on what the promotion apply (bcs i have 0 idea) choosing the least taxed item would result in the user paying the highest tax % possible (which they would not be happy about).
On the other hand doing that on the product with the highest tax % might be a good for the customer/Odoo user but seems rather fishy to me.

That's exactly what I "explained" through my previous use cases / screen shots?

The tax report is broken right now anyway (stable), since we didn't implement the tax on fixed discount, resulting in seller paying taxes to the administration he didn't even receive from the customer as taxes > amount total :facepalm: 
I don't think there is legislation in that regard, it was discussed a few years ago and I don't remember something about that.

The easiest way would be to follow the approach I suggested, sort lines by amount and by tax, to apply the discount on the biggest line first -> To be able for that line to "eat" all / most of the discount to avoid the split / line (or reduce it, if you have 5 products with 5 != taxes and SOL are 1$ 2$ 3$ 10$ 50$, the fixed discount is 55, sort it by amount will split the line only twice, if you don't sort by amount, it would split in 5 lines :grimacing: )

The reasoning is just to keep it simple and don't chose to favor the seller or the buyer but let the system detects how to apply the fixed discount to result in the least line splits / avoid line split.
From there, you can decide how "smart" you want to implement it (cf my previous message with use cases and 2 possibilities of implementation)

Let me know if something feels wrong.

File: addons/sale_coupon/models/sale_order.py (hunk 9 above)
Comment (pierre-pde) [reply addons/sale_coupon/models/sale_order.py:148]: HEllo, I confirm that I prefer Romain's suggestion.
Also, if the product generated by the program has a tax, we should use that one.
That way, admin can specify the tax they want on the product.

**Comment by pierre-pde** [comment]
@rdeodoo, @fdamhaut @dalaOdoo after discussion with Damien(DALA), we agreed that;
The flow suggested by Romain is the best functionally. 
To summarize it;
1. If the discount product has a tax defined for the relevant company --> use that tax.
2. Else, If all sales order line have the same tax --> use that tax (Note, this line doesn't need to be implemented in the algo. it will work naturally with #3 & #4)
3. Else, if the biggest sales.order.line.amount >= discount amount --> use tax of this sales.order.line
4. Else, do as easy as possible.. Probably split the tax amount with one line per different tax. & with amount proportional.

However, considering it might lots of code, we would agree to implement only 1 and 4, meaning;
1. If the discount product has a tax defined for the relevant company --> use that tax.
3. split the tax amount with one line per different tax. & with amount proportional.

This is up to you to be honest.
We agree that not having the discount splitted is better.
BUT, is it worth the supplementary lines of code considering that case might very rarely happen?

Also, being able to define the tax on the product is very important. e.g. if you think the discount shouldn't be taxable, you want to put a 0% tax on the product.

**Comment by rdeodoo** [comment]
Not sure how is "If there is a line able to absorb the whole discount, use that line tax" going to be "a lot more code" :shrug:

**Comment by dalaOdoo** [comment]
Be aware it has to be similar in PoS

**Comment by fdamhaut** [comment]
> 

@pierre-pde It's not much more code and It's already done :sweat_smile: as I guessed it would be the case
The actual implementation is slightly different but focus on the objective of having as few line as possible.

Instead of acting by line, I 'group' them by their taxes and apply the promotion on the tax group with the maximum possible amount (and the following if it's too much)

@rdeodoo mind giving a look at the implementation while I rewrite the tests :D ? I had to use `tax.compute_all()` to process the correct amount whatever the tax but that might be to expensive a call and we might want to say Fi if taxes are not %based and instead use the SO's lines to infer relation between price_unit and tax_included price

File: addons/sale_coupon/models/sale_order.py [hunk 12]
```diff
@@ -141,20 +141,81 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
         discount_amount = line.product_uom_qty * line.price_reduce * (program.discount_percentage / 100)
         return discount_amount
 
-    def _get_reward_values_discount(self, program):
-        if program.discount_type == 'fixed_amount':
-            taxes = program.discount_line_product_id.taxes_id
-            if self.fiscal_position_id:
-                taxes = self.fiscal_position_id.map_tax(taxes)
+    def _get_max_reward_values_per_tax(self, program, taxes):
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:144]: A bit misleading, I would expect to receive the reward values (as per the method name), but this is just about returning the total amount of all the lines (discount lines included) for the given taxes, except the lines of the current program.

File: addons/sale_coupon/models/sale_order.py [hunk 13]
```diff
@@ -185,8 +246,14 @@ def _get_reward_values_discount(self, program):
...
             # when processing lines we should not discount more than the order remaining total
             currently_discounted_amount = 0
+
+            tax_groups = set([line.tax_id for line in lines])
+            max_discount_per_tax_groups = {tax_ids: self._get_max_reward_values_per_tax(program, tax_ids) for tax_ids in tax_groups}
+
             for line in lines:
                 discount_line_amount = min(self._get_reward_values_discount_percentage_per_line(program, line), amount_total - currently_discounted_amount)
+                discount_line_amount = min(max_discount_per_tax_groups[line.tax_id], discount_line_amount)
+                max_discount_per_tax_groups[line.tax_id] -= discount_line_amount if discount_line_amount > 0 else 0
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:256]: Why do you need to alter the % discount mechanism?
How can a discount_line_amount (which will always be < than the SOL amount as it is a %age of it) be greater than the amount of multiple SOL including itself?
Is that because of the fixed discount amount?

File: addons/sale_coupon/models/sale_order.py [hunk 14]
```diff
@@ -141,20 +141,81 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
...
+
+        tax_groups = set([line.tax_id for line in lines])
+        max_discount_per_tax_groups = {tax_ids: self._get_max_reward_values_per_tax(program, tax_ids) for tax_ids in tax_groups}
+
+        for tax_ids in sorted(tax_groups, key=lambda tax_ids: max_discount_per_tax_groups[tax_ids], reverse=True):
+
+            if discount_amount <= 0:
+                return reward_lines.values()
+
+            curr_line = lines.filtered(lambda l: l.tax_id == tax_ids)
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:176]: confusing, `curr_line` is expected to hold a single line, rename it please :+1:

File: addons/sale_coupon/models/sale_order.py [hunk 15]
```diff
@@ -141,20 +141,81 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
...
+            lines_price = sum(curr_line.mapped(lambda l: l.price_reduce*l.product_uom_qty))
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:177]: Linter error, please install one

File: addons/sale_coupon/models/sale_order.py [hunk 16]
```diff
@@ -141,20 +141,81 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
...
             return [{
                 'name': _("Discount: ") + program.name,
                 'product_id': program.discount_line_product_id.id,
-                'price_unit': - self._get_reward_values_discount_fixed_amount(program),
+                'price_unit': -discount_amount,
+                'product_uom_qty': 1.0,
+                'product_uom': program.discount_line_product_id.uom_id.id,
+                'is_reward_line': True,
+                'tax_id': [(4, tax.id, False) for tax in program.discount_line_product_id.taxes_id],
+                }]
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:161]: linter error

File: addons/sale_coupon/models/sale_order.py [hunk 17]
```diff
@@ -141,20 +141,81 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
...
+
+    def _get_reward_values_percentage_amount(self, program):
+
+        # Invalidate multiline fixed_price discount line as they should apply after % discount
+        fixed_price_products = self._get_applied_programs().filtered(
+            lambda p: p.discount_type == 'fixed_amount' and p.discount_apply_on != 'specific_products'
+        ).mapped('discount_line_product_id')
+        self.order_line.filtered(lambda l: l.product_id in fixed_price_products).write({'price_unit': 0})
+
+
```
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:218]: linter error

File: addons/sale_coupon/tests/test_program_numbers.py [hunk 18]
```diff
@@ -1229,3 +1230,266 @@ def test_program_maximum_use_number_last_order(self):
...
+        order.recompute_coupon_lines()
+
+        self.assertEqual(order.amount_total, 5, 'Price should be 10$ - 5$(discount) = 5$')
+        self.assertEqual(order.amount_tax, 0, 'No taxes are applied yet')
+
+        sol.tax_id = self.tax_10pc_base_incl
+        order.recompute_coupon_lines()
+
+        self.assertEqual(order.amount_total, 5, 'Price should be 10$ - 5$(discount) = 5$')
+        self.assertEqual(float_compare(order.amount_tax, 5/11, precision_rounding=3), 0, '10% Tax included in 5$')
```
Comment (rdeodoo) [inline addons/sale_coupon/tests/test_program_numbers.py:1264]: Linter error

File: addons/sale_coupon/models/sale_order.py (hunk 8 above)
Comment (rdeodoo) [reply addons/sale_coupon/models/sale_order.py:459]: I really can't wrap my head around this change. Why the need of the `lines_to_keep` variable (that's the one I was speaking of, and which introduce some weird and hard to read conditions everywhere in this part of the code).

```py
                # Working on Case 3.
                # update_line update the line to the correct value and returns them if they should be unlinked
                update_to_remove = update_line(order, line, value)
                if not update_to_remove:
                    lines_to_keep += [(0, False, value)]
                    lines_to_remove -= line
                continue
        # Working on Case 2.
        if not value_found:
            lines_to_add += [(0, False, value)]
    # Case 3.
    line_update = []
    if lines_to_remove:
        line_update += [(3, line_id, 0) for line_id in lines_to_remove.ids]
        line_update += lines_to_keep
    line_update += lines_to_add
    order.write({'order_line': line_update})
```
=> Why does it need to be so over-complicated? Not saying this is wrong, but an explanation would be welcome.
```py
                # Working on Case 3.
                lines_to_remove -= line
                lines_to_remove += update_line(order, line, value)
                continue
        # Working on Case 2.
        if not value_found:
            lines_to_add += [(0, False, value)]
    # Case 3.
    line_update = [[(3, line_id, 0) for line_id in lines_to_remove.ids]
    line_update += lines_to_add
    order.write({'order_line': line_update})
```

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thanks for the changes.
Are you at the office tomorrow? We could try to meet to finally merge this one, it will be easier to avoid back & forth.

File: addons/sale_coupon/models/sale_order.py (hunk 13 above)
Comment (fdamhaut) [reply addons/sale_coupon/models/sale_order.py:256]: In case you apply a fixed amount before a % amount there is an issue without this check.
In most of the code, there is no issue as the fix force the application of % discount before fixed one but currently without this, if you applied a fixed price discount on a product and the a % which would make the price go lower than 0, you could have a negative total price which this line prevent.

After further thought, we could force the recomputation of discount line for specific product too and get rid of this.

**Comment by fdamhaut** [comment]
TY for the feedback :D

Unfortunatly I don't come on Tuesday but we could do that thursday.

As for the `lines_to_keep`, with the current implementation of SOL related to program, when you unlink 1 line related to a program you unlink all of them (https://github.com/odoo/odoo/blob/ed170550f79947b8946ad8efb233104bbcbd185a/addons/sale_coupon/models/sale_order.py#L493-L509) it does not cause error as any line remove will be recomputed later but this limit the creation of useless line when the old one where already correct.

That being said, I understand it's not evident and will add a comment :D

**Review by rdeodoo — APPROVED** [review]
@robodoo r+

Thank you for your work @fdamhaut

**Comment by Williambraecky** [comment]
I know I know
